node/core/pvf/common/src/worker.rs

```diff
@@ -82,6 +82,15 @@ pub fn worker_event_loop<F, Fut>(
 	F: FnMut(UnixStream) -> Fut,
 	Fut: futures::Future<Output = io::Result<Never>>,
 {
+	// Use `PR_SET_PDEATHSIG` to ensure that the child is sent a kill signal when the parent dies.
+	//
+	// NOTE: This technically has a race as the parent may have already died. In that case we will
+	// fail to read the socket later and just shutdown.
+	#[cfg(linux)]
+	if libc::prctl(libc::PR_SET_PDEATHSIG, libc::SIGKILL, 0, 0, 0) != 0 {
+		return
+	}
```

Comment:
```
Warning: the "parent" in this case is considered to be the
              thread that created this process.  In other words, the
              signal will be sent when that thread terminates (via, for
              example, [pthread_exit(3)](https://man7.org/linux/man-pages/man3/pthread_exit.3.html)), rather than after all of the
              threads in the parent process terminate.
```

Reply:
AKA, this isn't doing what you think.

Reply:
Good catch, hmm, I think since threads have their own pid this still addresses the race identified in this PR, I think.

```rs
// NOTE: On non-Linux this has a race condition between getting the pid and sending the
// signal -- the parent may have died and another process been assigned the same pid. On
// Linux this is not a problem -- we use `PR_SET_PDEATHSIG` to ensure that the child is sent
// a kill signal when the parent dies.
// [...]
let ppid = libc::getppid();
if ppid > 1 {
    libc::kill(ppid, libc::SIGTERM);
}
```

So the issue moves up a level, i.e. it becomes about avoiding leaked threads in the parent as opposed to leaked/orphaned child processes.

Reply:
Not sure which race you now mean, the one in the comment above or that we don't kill workers on node exit?

Reply:
I mean the race in the `NOTE` comment is addressed even with the caveat you posted from `man`. And as you pointed out threads will exit when the process does, so the caveat doesn't actually change anything. So I think this works.

Reply:
This is still somewhat equivalent, pvf-future is spawned on blocking pool, it can only be terminated along with a thread from the pool. If it happens, worker should be killed as well because pvf task is essential and shouldn't terminate. As long as it's documented, sounds ok to me.

Reply:
> This is still somewhat equivalent, pvf-future is spawned on blocking pool, it can only be terminated along with a thread from the pool.

Why do we spawn this on a blocking pool? 

But yeah it would need to be documented that this is a hard requirement now.

Reply:
Sorry @slumber I'm not really following - maybe my brain is running slow, can you explain more? I'm also not sure we should document implementation details too much, `PR_SET_PDEATHSIG` should work regardless of how the host is implemented.

I also removed some of the comments for now. Maybe documenting such rare corner cases was overkill, it just leads to noise.

Reply:
> Why do we spawn this on a blocking pool?

I honestly don't know.

> can you explain more?

Basti quoted documentation mentioning that your prctl call will terminate worker once parent **thread** terminates, not the node itself. Suppose pvf-future spawns worker and moves to a different tokio thread (tokio workers steal work). This thread terminating would mean that a worker would get killed, and, as Basti mentioned, you're not doing what you wanted to.

Then I noted that pvf future in fact is spawned on a blocking pool. Essentially, `tokio::spawn_blocking(block_on(pvf_host))`. Given this, your PR works. This thread terminating is equivalent of terminating pvf host, which is sufficient for worker termination. If we replace `spawn_blocking` with `spawn`, your PR doesn't work anymore, this should be documented.

Reply:
I see, you are correct. I'm not sure how I feel about relying on it being a blocking pool, since that fact can change. 

Does work-stealing mean that the thread that spawned work can actually die while the work is still active? That would surprise me since it seems unsound (precisely for cases like `PR_SET_PDEATHSIG`), but TBH I don't know how tokio's scheduler works under the hood.

Reply:
Work stealing means each worker thread has its own queue and they steal tasks from each other, "work stealing" is a term used by golang scheduler and tokio, you can read about it.

The main point is that task can move to another thread. For example, pvf future is on thread A and blocked, thread A steals another future and panics.